src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel_UpdateViewModelState.cs

```diff
@@ -0,0 +1,151 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Collections;
+using Microsoft.VisualStudio.Progression;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Newtonsoft.Json.Linq;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    internal partial class DocumentOutlineViewModel
+    {
+        /// <summary>
+        /// Queue for updating the state of the view model
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<ViewModelStateDataChange> _updateViewModelStateQueue;
+
+        public void EnqueueFilter(string newText)
+            => _updateViewModelStateQueue.AddWork(new ViewModelStateDataChange(newText, null, null, false));
+
+        public void EnqueueSelectTreeNode(CaretPosition caretPoint)
+            => _updateViewModelStateQueue.AddWork(new ViewModelStateDataChange(null, caretPoint, null, false));
+
+        public void EnqueueExpandOrCollapse(ExpansionOption option)
+            => _updateViewModelStateQueue.AddWork(new ViewModelStateDataChange(null, null, option, false));
+
+        /// <summary>
+        /// state that is only updated in <see cref="UpdateViewModelStateAsync"/>
+        /// </summary>
+        private string? _currentSearchText = null;
+        private CaretPosition? _currentCaretPosition = null;
+        private ExpansionOption? _currentExpansionOption = null;
```

Comment:
uh... hate this.  do we need mutable state like this?    this also needs doc'ing on threading constraints.